File: leetcode/binary_tree/sumNumbers.js
var sumNumbers = function(root) {
  function sumNumbersHelper(root) {
    if (root === null) return true;

    stack.push(root.val)

    const left = sumNumbersHelper(root.left);
    const right = sumNumbersHelper(root.right);

    if (left && right) sum += Number(stack.join(''));

    stack.pop()

    return false;
  }

  let sum = 0;
  const stack = [];
  sumNumbersHelper(root)

  return sum;
};

File: leetcode/binary_tree/sum_numbers.rb
def sum_number_helper(root, array, nums)
  return if root.nil?

  array << root.val
  left = sum_number_helper(root.left, array, nums)
  right = sum_number_helper(root.right, array, nums)

  if left.nil? && right.nil?
    nums << array.join.to_i
  end

  array.pop
end

def sum_numbers(root)
  array = []
  nums = []

  sum_number_helper(root, array, nums)

  nums.reduce(:+)
end
